src/core/task/tools/handlers/SearchFilesToolHandler.ts

```diff
@@ -0,0 +1,144 @@
+		const filePattern = block.params.file_pattern
+		const searchDescription = `'${regex}'${filePattern ? ` in '${filePattern}'` : ""}`
+
+		const sharedMessageProps = {
+			tool: "searchFiles",
+			path: getReadablePath(config.cwd, uiHelpers.removeClosingTag(block, "path", relPath)),
+			content: `Searching for ${searchDescription}`,
+			regex: uiHelpers.removeClosingTag(block, "regex", regex),
+			filePattern: filePattern ? uiHelpers.removeClosingTag(block, "file_pattern", filePattern) : undefined,
+			operationIsLocatedInWorkspace: await isLocatedInWorkspace(relPath),
+		}
+
+		const partialMessage = JSON.stringify(sharedMessageProps)
+
+		// Handle auto-approval vs manual approval for partial
+		if (await uiHelpers.shouldAutoApproveToolWithPath(block.name, relPath)) {
+			await uiHelpers.removeLastPartialMessageIfExistsWithType("ask", "tool")
+			await uiHelpers.say("tool", partialMessage, undefined, undefined, block.partial)
+		} else {
+			await uiHelpers.removeLastPartialMessageIfExistsWithType("say", "tool")
+			await uiHelpers.ask("tool", partialMessage, block.partial).catch(() => {})
+		}
+	}
+
+	async execute(config: TaskConfig, block: ToolUse): Promise<ToolResponse> {
+		// For partial blocks, return empty string to let coordinator handle UI
+		if (block.partial) {
+			return ""
+		}
+
+		const relDirPath: string | undefined = block.params.path
+		const regex: string | undefined = block.params.regex
+		const filePattern: string | undefined = block.params.file_pattern
+
+		// Validate required parameters
+		const pathValidation = this.validator.assertRequiredParams(block, "path")
+		if (!pathValidation.ok) {
+			config.taskState.consecutiveMistakeCount++
+			return await config.callbacks.sayAndCreateMissingParamError("search_files", "path")
+		}
+
+		if (!regex) {
+			config.taskState.consecutiveMistakeCount++
+			return await config.callbacks.sayAndCreateMissingParamError("search_files", "regex")
+		}
+
+		config.taskState.consecutiveMistakeCount = 0
+		const absolutePath = path.resolve(config.cwd, relDirPath!)
+
+		// Handle approval flow
+		const searchDescription = `'${regex}'${filePattern ? ` in '${filePattern}'` : ""}`
+		const sharedMessageProps = {
+			tool: "searchFiles",
+			path: getReadablePath(config.cwd, relDirPath!),
+			content: `Searching for ${searchDescription}`,
+			regex: regex,
+			filePattern: filePattern,
+			operationIsLocatedInWorkspace: await isLocatedInWorkspace(relDirPath!),
+		}
+
+		const completeMessage = JSON.stringify(sharedMessageProps)
+
+		if (await config.callbacks.shouldAutoApproveToolWithPath(block.name, relDirPath)) {
+			// Auto-approval flow
+			await config.callbacks.removeLastPartialMessageIfExistsWithType("ask", "tool")
+			await config.callbacks.say("tool", completeMessage, undefined, undefined, false)
+			config.taskState.consecutiveAutoApprovedRequestsCount++
+
+			// Capture telemetry
+			telemetryService.captureToolUsage(config.ulid, block.name, config.api.getModel().id, true, true)
```

Comment:
looks like we are capturing tool usage twice here since there s no early return that'd block the next  captureToolUsage